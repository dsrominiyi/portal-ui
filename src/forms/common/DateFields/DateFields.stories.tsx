import React from 'react';
import { storiesOf } from '@storybook/react';

import DateFields from '.';

const stories = storiesOf('DateFields', module);

const render = (values: DateFieldValues) => (props?: any) => (
  <DateFields values={values} onChange={props && props.onChange} />
);

export const fieldValues = { dd: '03', mm: '01', yyyy: '2020' };

export const storyRenders = {
  noValues: render({ dd: '', mm: '', yyyy: '' }),
  withValues: render(fieldValues)
};

stories.add('No values', storyRenders.noValues);
stories.add('With values', storyRenders.withValues);
