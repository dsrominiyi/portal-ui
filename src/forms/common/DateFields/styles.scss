@import '../../../style/vars';

.date-field {
  display: inline-flex;
  flex-direction: column;
  width: 8rem;
  margin-right: 5px;

  > div:first-child {
    color: $brand-stone-grey;
  }
}
