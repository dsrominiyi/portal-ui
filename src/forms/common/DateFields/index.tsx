import React, { FunctionComponent, Fragment } from 'react';

import TextField from '../../../components/TextField';

import { stripNonNumeric } from '../../../helpers/utilFuncs';

import styles from './styles.scss';

export interface Props {
  values: DateFieldValues;
  isInvalid?: boolean;
  testId?: string;
  onChange(values: DateFieldValues): void;
}

const addLeadingZero = (value: string) => (value.length === 1 ? `0${value}` : value);

export const DateFields: FunctionComponent<Props> = ({
  values: { dd, mm, yyyy },
  isInvalid,
  testId = '',
  onChange
}: Props) => {
  const updateDay = (value: string) => onChange({ dd: stripNonNumeric(value), mm, yyyy });
  const updateMonth = (value: string) => onChange({ dd, mm: stripNonNumeric(value), yyyy });
  const updateYear = (value: string) => onChange({ dd, mm, yyyy: stripNonNumeric(value) });

  return (
    <Fragment>
      <TextField
        label="Day"
        placeholder="DD"
        value={dd}
        maxLength={2}
        additionalClasses={[styles.dateField]}
        isInvalid={isInvalid}
        onChange={event => updateDay(event.target.value)}
        onBlur={event => updateDay(addLeadingZero(event.target.value))}
        attributes={{ 'test-id': testId ? `${testId}_day` : '' }}
      />
      <TextField
        label="Month"
        placeholder="MM"
        value={mm}
        maxLength={2}
        additionalClasses={[styles.dateField]}
        isInvalid={isInvalid}
        onChange={event => updateMonth(event.target.value)}
        onBlur={event => updateMonth(addLeadingZero(event.target.value))}
        attributes={{ 'test-id': testId ? `${testId}_month` : '' }}
      />
      <TextField
        label="Year"
        placeholder="YYYY"
        value={yyyy}
        maxLength={4}
        additionalClasses={[styles.dateField]}
        isInvalid={isInvalid}
        onChange={event => updateYear(event.target.value)}
        attributes={{ 'test-id': testId ? `${testId}_year` : '' }}
      />
    </Fragment>
  );
};

DateFields.displayName = 'DateFields';
export default DateFields;
