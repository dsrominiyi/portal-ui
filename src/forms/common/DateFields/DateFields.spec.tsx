import { shallow, ShallowWrapper } from 'enzyme';
import { fieldValues, storyRenders } from './DateFields.stories';

import testStorySnapshots from '../../../helpers/test/testStorySnapshots';

let component: ShallowWrapper;
const onChange = jest.fn();

describe('DateFields', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    component = shallow(storyRenders.withValues({ onChange }));
  });

  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders);
  });

  test('changing the day field calls onChange with updated field values', () => {
    const { mm, yyyy } = fieldValues;
    const dd = '22';
    const event = { target: { value: dd } };
    component
      .find('TextField')
      .at(0)
      .simulate('change', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm, yyyy });
  });

  test('changing the month field calls onChange with updated field values', () => {
    const { dd, yyyy } = fieldValues;
    const mm = '07';
    const event = { target: { value: mm } };
    component
      .find('TextField')
      .at(1)
      .simulate('change', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm, yyyy });
  });

  test('changing the year field calls onChange with updated field values', () => {
    const { dd, mm } = fieldValues;
    const yyyy = '1996';
    const event = { target: { value: yyyy } };
    component
      .find('TextField')
      .at(2)
      .simulate('change', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm, yyyy });
  });

  test('a leading 0 is added to a single digit day value on blur from the field', () => {
    const { mm, yyyy } = fieldValues;
    const dd = '2';
    const event = { target: { value: dd } };
    component
      .find('TextField')
      .at(0)
      .simulate('blur', event);

    expect(onChange).toHaveBeenCalledWith({ dd: '02', mm, yyyy });
  });

  test('a leading 0 is added to a single digit month value on blur from the field', () => {
    const { dd, yyyy } = fieldValues;
    const mm = '7';
    const event = { target: { value: mm } };
    component
      .find('TextField')
      .at(1)
      .simulate('blur', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm: '07', yyyy });
  });

  test('no change is made on blur from the field if the value is 2 digits', () => {
    const { dd, mm, yyyy } = fieldValues;
    const event = { target: { value: dd } };
    component
      .find('TextField')
      .at(0)
      .simulate('blur', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm, yyyy });
  });

  test('non numeric values are stripped from the day field', () => {
    const { mm, yyyy } = fieldValues;
    const dd = '2b';
    const event = { target: { value: dd } };
    component
      .find('TextField')
      .at(0)
      .simulate('change', event);

    expect(onChange).toHaveBeenCalledWith({ dd: '2', mm, yyyy });
  });

  test('non numeric values are stripped from the month field', () => {
    const { dd, yyyy } = fieldValues;
    const mm = '7Â£';
    const event = { target: { value: mm } };
    component
      .find('TextField')
      .at(1)
      .simulate('change', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm: '7', yyyy });
  });

  test('non numeric values are stripped from the year field', () => {
    const { dd, mm } = fieldValues;
    const yyyy = '199K';
    const event = { target: { value: yyyy } };
    component
      .find('TextField')
      .at(2)
      .simulate('change', event);

    expect(onChange).toHaveBeenCalledWith({ dd, mm, yyyy: '199' });
  });
});
