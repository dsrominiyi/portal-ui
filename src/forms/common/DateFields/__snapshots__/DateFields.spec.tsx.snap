// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateFields component story renders match snapshots: noValues 1`] = `
<Fragment>
  <TextField
    additionalClasses={
      Array [
        "dateField",
      ]
    }
    attributes={
      Object {
        "test-id": "",
      }
    }
    label="Day"
    maxLength={2}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="DD"
    value=""
  />
  <TextField
    additionalClasses={
      Array [
        "dateField",
      ]
    }
    attributes={
      Object {
        "test-id": "",
      }
    }
    label="Month"
    maxLength={2}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="MM"
    value=""
  />
  <TextField
    additionalClasses={
      Array [
        "dateField",
      ]
    }
    attributes={
      Object {
        "test-id": "",
      }
    }
    label="Year"
    maxLength={4}
    onChange={[Function]}
    placeholder="YYYY"
    value=""
  />
</Fragment>
`;

exports[`DateFields component story renders match snapshots: withValues 1`] = `
<Fragment>
  <TextField
    additionalClasses={
      Array [
        "dateField",
      ]
    }
    attributes={
      Object {
        "test-id": "",
      }
    }
    label="Day"
    maxLength={2}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="DD"
    value="03"
  />
  <TextField
    additionalClasses={
      Array [
        "dateField",
      ]
    }
    attributes={
      Object {
        "test-id": "",
      }
    }
    label="Month"
    maxLength={2}
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="MM"
    value="01"
  />
  <TextField
    additionalClasses={
      Array [
        "dateField",
      ]
    }
    attributes={
      Object {
        "test-id": "",
      }
    }
    label="Year"
    maxLength={4}
    onChange={[Function]}
    placeholder="YYYY"
    value="2020"
  />
</Fragment>
`;
